### DISCUSSION TITLE
Send data to client without get a request (unsolicited)

### DISCUSSION THREAD

**Discussion by majidnow** [uNetworking/uWebSockets#1147] [Q&A]
Hi
How can I send data to client without request?

**Comment by NikitaMishin** [comment]
same question

**Comment by Jacob-Burckhardt** [comment]
Here's an example of how to do that.  The only "send" call in the below code is in a timer function.  That means it sends data based on a timer instead of based on receiving something from the client.  

The timer function runs in the same thread as the uWebSockets loop which means it is thread-safe.

```
#include "App.h"
   
using namespace std;

struct PerSocketData {};

uWS::WebSocket<false, true, PerSocketData> *gws=nullptr;
    
int main() {
   auto loop = uWS::Loop::get();

   struct us_timer_t *delayTimer = us_create_timer((struct us_loop_t *) loop, 0, 0);

   us_timer_set(delayTimer, [](struct us_timer_t *) {
                               if (gws) {
                                  cout << "calling send" << endl;
                                  gws->send("from server", uWS::OpCode::TEXT);
                               }
                            }, 1000, 1000);
   
   uWS::App app;

   app.ws<PerSocketData>("/*", {
         .idleTimeout = 0,
         .sendPingsAutomatically = false,
         .open = [](auto *ws) {
                    gws = ws;
                 },
         .close = [](auto */*ws*/, int /*code*/, std::string_view /*message*/) {
                     gws = nullptr;
                  }
      }).listen(9001, [](auto *) {
                      });

   app.run();
}

```

**Reply by NikitaMishin** [reply]
Did a little bit diff--- used either postCb or preCb for invoking Cb where the thread pop from custom event queue. 
Thanks for response!

**Reply by NikitaMishin** [reply]
But with your and mine approach the actual sending is postponed. I suppose for mostly all wss clients this is an issue. Am i right?

**Reply by Jacob-Burckhardt** [reply]
I ran a test to measure the time.  I made my uWebSockets server send to a WebSocket++ client with both running on the same host.  The delay was 59 to 227 microseconds.  Some of that delay is caused by the OS and the WebSocket++ library which means uWebSockets delay is even less than the number I gave.  Those numbers seem very low to me, so I don't think you have to worry about delays.

I measured from when I called clock_gettime on the server to when I received the data on the client and called clock_gettime on the client as well to compare the times.  My server puts the time from clock_gettime into a binary string and calls send in the uWebSockets timer callback.  

However, the timer callback is called only once every 1 second in the above example.  But that's just so that you won't be overwhelmed with fast output.  In my production code, I call it every 50 milliseconds and I have seen no indication of excessive delay in production.  You can experiment with lower timer periods in my example and see how fast you can get it without using excessive CPU time.

Instead of a timer you can call defer so that the uWebSockets main loop will send it as soon as it is done with whatever else it is doing.

I'm unfamiliar with postCb and preCb so I can't say how responsive that would be.
